syft/file/digest_cataloger.go

```diff
@@ -74,16 +74,18 @@ func (i *DigestsCataloger) catalogLocation(resolver source.FileResolver, locatio
 		return nil, fmt.Errorf("unable to observe contents of %+v: %+v", location.RealPath, err)
 	}
 
+	if size == 0 {
+		return make([]Digest, 0), nil
+	}
+
 	result := make([]Digest, len(i.hashes))
-	if size > 0 {
-		// only capture digests when there is content. It is important to do this based on SIZE and not
-		// FILE TYPE. The reasoning is that it is possible for a tar to be crafted with a header-only
-		// file type but a body is still allowed.
-		for idx, hasher := range hashers {
-			result[idx] = Digest{
-				Algorithm: cleanAlgorithmName(i.hashes[idx].String()),
-				Value:     fmt.Sprintf("%+x", hasher.Sum(nil)),
-			}
+	// only capture digests when there is content. It is important to do this based on SIZE and not
+	// FILE TYPE. The reasoning is that it is possible for a tar to be crafted with a header-only
+	// file type but a body is still allowed.
+	for idx, hasher := range hashers {
+		result[idx] = Digest{
+			Algorithm: cleanAlgorithmName(i.hashes[idx].String()),
```

Comment:
Optional but could help — It looks like in most calls to `cleanAlgorithmName`, we have a value but are calling `.String()` on that value just in time. Usually when I notice this, I realize I should at minimum consider changing the parameter type from `string` to `fmt.Stringer`, to clean up the consuming code and consolidate the extra `.String()` work in the called function. In this particular case though, it looks like we could possibly go even further and take a parameter of type `crypto.Hash` instead. This would immediately simplify 3 out of the 4 calls being made to `cleanAlgorithmName`. The 4th call has a string already, because `NewDigestsCataloger` is taking a slice of strings. But it might be advantageous to modify this too to use `crypto.Hash` to ensure usable hash algorithms are supplied. If conversion from a string is needed because of user configuration, that should probably be pushed away from the application logic toward the user boundary as much as possible.

Reply:
I agree with the overall logic of moving the string work to the caller of NewDigestsCataloger, but I think this is a bit of a trade off since the `cleanAlgorithmName` is used for a couple purposes and will now need to be exported for both purposes (parsing strings and cleaning hash names).